### PR TITLE
Add syntax for higher order quote pattern holes

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Make typer explicitly type higher-order quasiquotes
* Add higher-order quasi-pattern concept to the Matcher (avoid ambiguities with splice applied to arguments)
* Introduce unambiguous syntax for higher-order quasi-patterns
* Add documentation

```scala
case ${ val a = 3; $body(a): Int } => // higher-order quasi-patterns
case ${ val a = 3; ${fn}(a): Int } => // match an application 
```

### CODE REVIEW

**Comment by nicolasstucki** [comment]
As discussed offline, we will try to encode it with the syntax
```scala
case ${ val a = 3; ${body}{a}: Int } => 
```

**Comment by nicolasstucki** [comment]
Note that there is a possible improvement for multiple parameters if we allow commas in block in the parser `${f}{x,y,z}`.

**Comment by odersky** [comment]
I am still  not sure about the best syntax. My original proposal was to use

```scala
$body{y}
```
for the instantiated splice, and to reserve `${body}{y}` is body should be spliced as is and then be applied to the block `{y}`. But it gets ugly for multiple parameters. I am a fan of neither of the alternatives
```scala
$body{y, z}
$body{(y, z)}
```
The first uses a comma in a block, which never happens otherwise and the second looks like `(y, z)` is a tupled argument where it is not. If we have to choose, I'd go for the first. 

But before settling on this, a question: If we leave it as
```scala
$block(y)
```
how often (relatively speaking) is a higher-order hole? If that's the common case, can we maybe live with the fallback
```scala
${block}(y)
```
for the uncommon case where $block is spliced in front of the literal argument list `(y)`?

**Comment by nicolasstucki** [comment]
@odersky I updated the syntax to the one we discussed in the weekly meeting.

File: compiler/src/dotty/tools/dotc/ast/untpd.scala [hunk 1]
```diff
@@ -103,7 +103,9 @@ object untpd extends Trees.Instance[Untyped] with UntypedTreeInfo {
   }
   case class Throw(expr: Tree)(implicit @constructorOnly src: SourceFile) extends TermTree
   case class Quote(quoted: Tree)(implicit @constructorOnly src: SourceFile) extends TermTree
-  case class Splice(expr: Tree)(implicit @constructorOnly src: SourceFile) extends TermTree
+  case class Splice(expr: Tree)(implicit @constructorOnly src: SourceFile) extends TermTree {
+    def isInBrackets: Boolean = span.end != expr.span.end
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/ast/untpd.scala:107]: I think that should be `inBraces`

File: docs/docs/reference/metaprogramming/macros.md [hunk 2]
```diff
@@ -744,5 +744,51 @@ trait Show[-T] {
...
+
+```scala
+eval { // expands to the code: (16: Int)
+  val x: Int = 4
+  x * x
+}
+```
+
+We can also close over several bindings using `$b(a1, a2, ..., an)`.
+To match an actual application we can use brackets on the function part `${b}(a1, a2, ..., an)`.
```
Comment (odersky) [inline docs/docs/reference/metaprogramming/macros.md:790]: braces on

**Pull request merged**